sound/soc/sof/topology.c

```diff
@@ -1693,16 +1693,22 @@ int sof_load_pipeline_ipc(struct device *dev,
 			  struct sof_ipc_comp_reply *r)
 {
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
-	int ret = sof_core_enable(sdev, pipeline->core);
+	int ret;
+
+	mutex_lock(&sdev->cores_status_mutex);
 
+	ret = sof_core_enable(sdev, pipeline->core);
 	if (ret < 0)
-		return ret;
+		goto done;
 
 	ret = sof_ipc_tx_message(sdev->ipc, pipeline->hdr.cmd, pipeline,
 				 sizeof(*pipeline), r, sizeof(*r));
```

Comment:
> `sof_ipc_tx_message()` takes a mutex internally, nesting (different) mutexes can be done of course, but should rather be avoided if possible. Do we really need to include it under the mutex too? Wouldn't it be enough to use the mutex inside `sof_core_enable()`?

do you mean the ipc->tx_mutex? we have different mutexes for different resources access, using nesting mutexes should be fine in case we can make sure there is no dead lock possibility. Here we could in scenario of holding cores_status_mutex and asking for ipc->tx_mutex, but no opposite scenario existed, so I think we are fine.

Reply:
This should really be a
```c
sof_core_get(sdev, pipeline->core);
```
where sof_core_get/put would do the ref counting and enable/disable.